packages/react-components/react-positioning/src/middleware/maxSize.ts

```diff
@@ -6,35 +6,105 @@ export interface MaxSizeMiddlewareOptions extends Pick<PositioningOptions, 'over
   container: HTMLElement | null;
 }
 
+/**
+ * AutoSizes contains many options from historic implementation.
+ * Now options 'always'/'height-always'/'width-always' are obsolete.
+ * This function maps them to true/'height'/'width'
+ */
+const normalizeAutoSize = (
+  autoSize?: PositioningOptions['autoSize'],
+): { applyMaxWidth: boolean; applyMaxHeight: boolean } => {
+  switch (autoSize) {
+    case 'always':
+    case true:
+      return {
+        applyMaxWidth: true,
+        applyMaxHeight: true,
+      };
+
+    case 'width-always':
+    case 'width':
+      return {
+        applyMaxWidth: true,
+        applyMaxHeight: false,
+      };
+
+    case 'height-always':
+    case 'height':
+      return {
+        applyMaxWidth: false,
+        applyMaxHeight: true,
+      };
+
+    default:
+      return {
+        applyMaxWidth: false,
+        applyMaxHeight: false,
+      };
+  }
+};
+
+/**
+ * floating-ui `size` middleware uses floating element's height/width to calculate available height/width.
+ * This middleware only runs once per lifecycle, resetting styles applied by maxSize from previous lifecycle.
+ * Then floating element's original size is restored and `size` middleware can calculate available height/width correctly.
+ */
+export const resetMaxSize = (autoSize: PositioningOptions['autoSize']): Middleware => ({
+  name: 'resetMaxSize',
+  fn({ middlewareData: { maxSizeAlreadyReset }, elements }) {
+    if (maxSizeAlreadyReset) {
+      return {};
+    }
+
+    const { applyMaxWidth, applyMaxHeight } = normalizeAutoSize(autoSize);
```

Comment:
Not sure if it makes sense, but you can normalize during configuration step and pass normalized values there. The small benefit is that you will do it once for both `maxSize` & `resetMaxSize`